pkg/images/file/file.go

```
@@ -0,0 +1,54 @@
+/*
+Copyright 2023 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package file
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"gopkg.in/yaml.v3"
+)
+
+// ImageProducer is a file-based image producer.
+// The expected file format is a YAML file like the following:
+// <go-package-ref>: <image-ref>
+//
+// For example:
+//
+// knative.dev/reconciler-test/cmd/eventshub: quay.io/myregistry/eventshub
+// # ... other images ...
+func ImageProducer(filepath string) func(ctx context.Context, pack string) (string, error) {
+	images := make(map[string]string, 2 /* */)
+
+	bytes, err := os.ReadFile(filepath)
+	if err != nil {
+		panic("failed to read file " + filepath + ": " + err.Error())
+	}
+
+	if err := yaml.Unmarshal(bytes, images); err != nil {
+		panic("failed to unmarshal images file: " + err.Error() + "\nContent:\n" + string(bytes))
+	}
+
+	return func(ctx context.Context, pack string) (string, error) {
+		im, ok := images[pack]
+		if !ok {
+			return "", fmt.Errorf("image for package %s not found in %+v", pack, images)
```

Do we want to return an error? This means we would have to "override" all images IIUC. It would be great if it would be allowed to override only certain images, so e.g. some `ko://...` images could be still in place :shrug:

---

Wouldn't that produce an empty image and so an error when deploying the resource?

---

or are you thinking that instead of having one image producer `xor` the other, we would need to compose image producers

---

> or are you thinking that instead of having one image producer xor the other, we would need to compose image producers?

more something like this. So in case it is not found in the defined image producer, it falls back to the default. But not sure if this is kind of unexpected/intransparent for the user... But it would allow to not have to override all used images

---

Yeah, that's why better to be explicit but also I don't have the use case of using both pre-built images and ko, so I'd leave that fallback behavior for now